devicetypes/smartthings/2015-samsung-smart-tv.src/2015-samsung-smart-tv.groovy

```diff
@@ -0,0 +1,140 @@
+        
+        command "volumeUp"
+		command "volumeDown"
+	}
+
+	tiles {
+        standardTile("switch", "device.switch", width:1, height:1) {
+        	state "on", label:'${name}', action: "switch.off", icon: "st.Electronics.electronics15", backgroundColor: "#3A935A"
+			state "off", label:'${name}', action: "switch.on", icon: "st.Electronics.electronics15", backgroundColor: "#ffffff"
+		}
+
+		valueTile("volume", "device.volume", width: 1, height: 1) {
+            state "volume", label:'${currentValue}\nvol', backgroundColor: "#1C808A"
+    	}
+        
+		standardTile("volumeUp", "device.button", width:1, height:1, decoration:"flat") {
+            state "default", label:"vol", action:"volumeUp", icon:"st.custom.buttons.add-icon"
+        }
+
+		standardTile("volumeDown", "device.button", width:1, height:1, decoration:"flat") {
+            state "default", label:"vol", action:"volumeDown", icon:"st.custom.buttons.subtract-icon"
+        }
+
+        standardTile("refresh", "device.refresh", width:1, height:1, decoration: "flat") {
+			state "default", label:"", action:"refresh.refresh", icon:"st.secondary.refresh"
+		}
+      
+        main(["switch"])
+
+        details(["switch", "volumeUp", "volumeDown", "refresh"])	}
+}
+
+def parse(String description) {
+	def parsedEvent = parseLanMessage(description)
+    def results
+    if(parsedEvent.body) {
+    	def type = parsedEvent.headers."Content-Type"
+		if(type?.contains("json"))
+        {
+        	def parsedBody = parseJson(parsedEvent.body)
+			results = createEvent(name: "${parsedBody.type}", value: "${parsedBody.value}")
+        }
+    }
+    results
+}
+
+def on() {
+    // To Do : make SmartThings Hub generate magic packet to turn on TV
+}
+
+def off() {
+    sendRequest("set", "power", "off")
+}
+
+def volumeUp() {
+	sendRequest("set", "volume", "up")
+}
+
+def volumeDown() {
+	sendRequest("set", "volume", "down")
+}
+
+def refresh() {
+	String mac = device.hub.getDataValue("macAddress")
+    mac = mac.toLowerCase()
+	sendRequest("get", "all", mac)
+}
+
+def sendRequest( String request, String type, String value ) {
+    String body
+    if( type == "notification" || type == "camera" )
+    {
+    	body = "{\"request\":\"" + request + "\",\"type\":\"" + type + "\",\"value\":" + value + "}"
+    }
+    else
+    {
+    	body = "{\"request\":\"" + request + "\",\"type\":\"" + type + "\",\"value\":\"" + value + "\"}"
+    }
+    
+	def httpRequest = [
+      	method:		"POST",
+        path: 		"/smartthings/tvfeature",
+        body : 		"${body}\n",
+
+        headers:[
+        			Host: getHostAddress(),
+                    /Content-Type/: "application/json",
+        		],
+        ]
+    
+	def hubAction = new physicalgraph.device.HubAction(httpRequest)
+    return hubAction
+}
+
+private getHostAddress() {
+    def ip = getDataValue("ip")
+    def port = getDataValue("port")
+
+    if (!ip || !port) {
+        def parts = device.deviceNetworkId.split(":")
```

Comment:
The Device Network ID (DNI) should not be an IP address. Since this implementation is not migrating existing devices installed with an improper DNI, this is not necessary and should be removed to avoid confusion.